### PR TITLE
export all unicode helper functions

### PR DESCRIPTION
**Pull request by ahimberg**
We were exporting these functions via .def files instead of dllexport.

Update to use the macro and some vcxproj changes for desktop to reduce warnings about duplicate functions in unit/integration tests.

I did not use the API_() form since that changes the calling convention, which might be fine but is slightly annoying for copy in a dll ABI compat

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2825)

### CODE REVIEW

File: vnext/include/ReactWindowsCore/unicode.h [hunk 1]
```diff
@@ -50,11 +52,11 @@ class UnicodeConversionException : public std::runtime_error {
 //
 // For (2), utf8 must be null terminated. The behavior is undefined otherwise.
 //
-/* (1) */ std::wstring utf8ToUtf16(const char *utf8, size_t utf8Len);
-/* (2) */ std::wstring utf8ToUtf16(const char *utf8);
-/* (3) */ std::wstring utf8ToUtf16(const std::string &utf8);
+/* (1) */ REACTWINDOWS_EXPORT std::wstring utf8ToUtf16(const char *utf8, size_t utf8Len);
```
Comment (aeulitz) [inline vnext/include/ReactWindowsCore/unicode.h]: >REACTWINDOWS_EXPORT [](start = 10, length = 19)

Why is that change needed? We're trying to introduce a "safe" ABI for react-native-win32. On first glance, this change appears to counter our efforts (regression in the sense that it seems to increase the number of exported functions).

Comment (ahimberg) [reply vnext/include/ReactWindowsCore/unicode.h]: We're consuming the nuget package for uwp and trying to use utf8to16, none of the functions were exported and giving them linker errors using the header.

do you have documentation on how you are exporting things now for win32?  I can find a way to make it uwp specific but I don't want to put more mangled strings into the .def files

**Comment by hansenyy** [comment]
@msftbot Make sure @aeulitz sign off on this pull request.

File: vnext/include/ReactWindowsCore/unicode.h (hunk 1 above)
Comment (aeulitz) [reply vnext/include/ReactWindowsCore/unicode.h]: We're planning on using C++/WinRT to define the Win32 ABI (which AFAIK should be similar/the same as the ABI of UWP components). No docs yet, only a few examples. I would like to set up a "review board" of folks designing/monitoring the RN API/ABI. Would you be interested in joining this?

From a separation-of-concerns perspective: Does RN really need to export string utilities? Wouldn't a better API-design for RN consist of endpoint methods with well-defined string args/return values?

---
In reply to: [307068623](https://github.com/microsoft/react-native-windows/pull/2825#discussion_r307068623) [](ancestors = 307068623)

Comment (ahimberg) [reply vnext/include/ReactWindowsCore/unicode.h]: That a bit unclear if utf conversions should be exported.  Our big office apps have string helper apis for this stuff already - but in places where we've written apps that are not depending on office, we don't have those and using consistent code with the internals seemed better than copy pasting over and over.  It's also taken awhile with a lot of bugs to have stable code here so just writing your own is likely to introduce those same bugs in some other codepath.  Who needs these conversions?  well with UWP native modules (and even reka) you end up pretty much guaranteed because the JS api boundary is utf8 but system APIs all require utf16.

Comment (aeulitz) [reply vnext/include/ReactWindowsCore/unicode.h]: I set up GitHub code owners on the .def files so that we monitor changes to the RN API/ABI and prevent regression with respect to our safe ABI efforts. As laudable as the switch to the REACTWINDOWS_EXPORT macro is, it makes policing the API/ABI harder. Would it be possible for you to limit the effect of REACTWINDOWS_EXPORT to the UWP platform, and to revert your changes to the *-win32.*.def files?

---
In reply to: [307085023](https://github.com/microsoft/react-native-windows/pull/2825#discussion_r307085023) [](ancestors = 307085023)
